Realm/ObjectStore/object_store.cpp

```diff
@@ -0,0 +1,418 @@
+            changed = true;
+        }
+
+        // keep track of names to figure out what tables need deletion
+        target_type_names.insert(object_schema.name);
+    }
+
+    // second pass adds/removes columns for out of date tables
+    for (auto& target_object_schema : to_update) {
+        TableRef table = table_for_object_type(group, target_object_schema->name);
+        ObjectSchema current_schema(group, target_object_schema->name, table.get());
+        vector<Property> &target_props = target_object_schema->properties;
+
+        // add missing columns
+        for (auto& target_prop : target_props) {
+            auto current_prop = current_schema.property_for_name(target_prop.name);
+
+            // add any new properties (new name or different type)
+            if (!current_prop || property_has_changed(*current_prop, target_prop)) {
+                switch (target_prop.type) {
+                        // for objects and arrays, we have to specify target table
+                    case PropertyTypeObject:
+                    case PropertyTypeArray: {
+                        realm::TableRef link_table = ObjectStore::table_for_object_type(group, target_prop.object_type);
+                        target_prop.table_column = table->add_column_link(realm::DataType(target_prop.type), target_prop.name, *link_table);
+                        break;
+                    }
+                    default:
+                        target_prop.table_column = table->add_column(realm::DataType(target_prop.type), target_prop.name);
+                        break;
+                }
+                changed = true;
+            }
+        }
+
+        // remove extra columns
+        sort(begin(current_schema.properties), end(current_schema.properties), [](Property &i, Property &j) {
+            return j.table_column < i.table_column;
+        });
+        for (auto& current_prop : current_schema.properties) {
+            auto target_prop = target_object_schema->property_for_name(current_prop.name);
+            if (!target_prop || property_has_changed(current_prop, *target_prop)) {
+                table->remove_column(current_prop.table_column);
+                changed = true;
+            }
+        }
+
+        // update table metadata
+        if (target_object_schema->primary_key.length()) {
+            // if there is a primary key set, check if it is the same as the old key
+            if (current_schema.primary_key != target_object_schema->primary_key) {
+                set_primary_key_for_object(group, target_object_schema->name, target_object_schema->primary_key);
+                changed = true;
+            }
+        }
+        else if (current_schema.primary_key.length()) {
+            // there is no primary key, so if there was one nil out
+            set_primary_key_for_object(group, target_object_schema->name, "");
+            changed = true;
+        }
+    }
+
+    // remove primary key entries for deleted types
```

Comment:
This removes the primary key for tables which still exist that are not present in the object schema., which is not even remotely the same thing as removing the primary key when a table is removed.  Even if it did the more sensible thing of removing entries in the pk table which mention tables that don't exist, this runs before the user gets the chance to call `deleteDataForClassName:`, so the stale data will persist until the next time the schema is initialized (and thus require an extra write transaction then).